File: templates/ruby/bot.rb
#!/usr/bin/env ruby -w

require 'io/console'
require 'socket'

def read_view()
	view = $socket.gets
	# we know the view has as many lines as columns minus
	# the trailing line feed
	for _ in 3..view.length do
		view += $socket.gets
	end
	view
end

$socket = TCPSocket.open(ARGV[0] || 'localhost', ARGV[1] || 63187)

while true
	view = read_view
	print view
	print "Command (q<>^v): "
	STDOUT.flush
	cmd = STDIN.getch
	puts
	case cmd
	when 'q'
		break
	end
	$socket.putc cmd
end

$socket.close

File: templates/perl/bot.pl
#!/usr/bin/env perl

use strict;
use warnings;
use Socket;

my $remote_host = $ARGV[0] || 'localhost';
my $remote_port = $ARGV[1] || 63187;

socket(SOCKET, PF_INET, SOCK_STREAM, getprotobyname('tcp'));

my $internet_addr = inet_aton($remote_host)
	or die "Couldn't convert $remote_host into an Internet address: $!\n";

my $paddr = sockaddr_in($remote_port, $internet_addr);

connect(SOCKET, $paddr)
	or die "Couldn't connect to $remote_host:$remote_port : $!\n";

sub read_view {
	my $view = <SOCKET>;
	my $lines = length $view;
	# we know the view has as many lines as columns minus
	# the trailing line feed
	while (--$lines > 1) {
		$view .= <SOCKET>;
	}
	return $view;
}

while (1) {
	my $view = read_view() or last;
	print $view;
	print 'Command (q<>^v): ';
	my $cmd = <>;
	$cmd = substr($cmd, 0, 1);
	if ($cmd eq 'q') {
		last;
	} elsif ($cmd eq "\n") {
		$cmd = '^'
	}
	send(SOCKET, $cmd, 0);
}

close(SOCKET);
